### PR TITLE
[Timeline] Use dpv population as per capita denominator

### PR DESCRIPTION
**Pull request by intrepiditee**
Features
- The first denominator in the list is passed to to the chart
- When grouped together in one chart, different StatVars have their own denominators
- The denominator is stored in the URL with the node path
- An asterisk is added next to the per capita button that redirects to the FAQ page

Screen Shot 2020-10-06 at 4 16 06 PM

Issues
- Denominators are not filtered
- We can consider also modifying `statsvar_path.json` to include the denominators with the node path so that if the denominator is missing from the URl, we can still load it

### CODE REVIEW

File: static/js/tools/timeline_page.tsx [hunk 1]
```diff
@@ -246,6 +252,12 @@ class Page extends Component<Record<string, unknown>, PageStateType> {
                     chartOptions={this.state.chartOptions}
                     setPC={this.setChartPerCapita.bind(this)}
                     initialPC={this.params.allPerCapita}
+                    denominators={Object.entries(
+                      this.state.statsVarNodes
+                    ).reduce((denominators, [dcid, node]) => {
+                      denominators[dcid] = node.denominators;
+                      return denominators;
+                    }, {})}
```
Comment (beets) [inline static/js/tools/timeline_page.tsx:260]: this is neat!

File: static/js/tools/timeline_util.ts [hunk 2]
```diff
@@ -247,19 +266,31 @@ class TimelineParams {
...
           } else {
-            this.addStatsVar(statsVarInfo[0], []);
+            this.addStatsVar(dcid, [], []);
           }
         } else {
-          this.addStatsVar(statsVarInfo[0], statsVarInfo.splice(1));
+          const pathsAndDenominators = statsVarInfo.splice(1);
+          this.addStatsVar(
+            dcid,
+            pathsAndDenominators.filter(
```
Comment (beets) [inline static/js/tools/timeline_util.ts:287]: can you describe these magic numbers? in particular, why 9 (it's hard to follow the format of the json, so would be good to add some documentation here. is there a better var name for elem? and should those be numbers instead of strings?)

**Review by beets — COMMENTED** [review]
thanks! this looks really complete, though have a small question below

File: server/webdriver_tests/screenshot/screenshot_test.py [hunk 3]
```diff
@@ -115,6 +115,9 @@ def test_pages_and_sreenshot(self):
             charts = self.driver.find_elements_by_class_name(
                 test_info['test_class'])
             # Assert there are charts.
+            if not len(charts):
```
Comment (shifucun) [inline server/webdriver_tests/screenshot/screenshot_test.py:118]: remove log?

File: static/data/hierarchy_statsvar.json [hunk 4]
```diff
@@ -127217,7 +127217,8 @@
      }
     ]
    }
-  ]
```
Comment (shifucun) [inline static/data/hierarchy_statsvar.json:127220]: there is no pv tree code change, are these from the run of the latest pv tree tool?

File: static/js/tools/statsvar_menu.tsx [hunk 5]
```diff
@@ -37,10 +38,15 @@ interface NodePropType {
...
+  d?: string[]; // possible denominators
   selectedNodes: StatsVarNode; // path to all the selected statsVars
   nodePath: string[]; // path to current node
   addStatsVarTitle: (statsVarId: string, statsVarName: string) => void; // pass the title of selected statsVar to parent
-  addStatsVar: (statsVar: string, nodePath: string[]) => void; // function for adding statsVar
+  addStatsVar: (
+    statsVar: string,
+    nodePath: string[],
+    denominators: string[]
+  ) => void; // function for adding statsVar
```
Comment (shifucun) [inline static/js/tools/statsvar_menu.tsx:49]: can you check if this affects other client like choropleth tool which also uses the menu

File: static/js/tools/timeline_chart.tsx [hunk 6]
```diff
@@ -98,11 +100,14 @@ class Chart extends Component<ChartPropsType, unknown> {
               this.props.setPC(this.props.groupId, !this.props.perCapita);
             }}
           ></button>
+          <a href="/faq">
+            <span>&#42;</span>
```
Comment (shifucun) [inline static/js/tools/timeline_chart.tsx:104]: can use {*} in React

**Review by shifucun — COMMENTED** [review]
looks pretty good, some minor comments

File: static/js/tools/timeline_page.tsx (hunk 1 above)
Comment (intrepiditee) [reply static/js/tools/timeline_page.tsx:260]: thanks

File: static/data/hierarchy_statsvar.json (hunk 4 above)
Comment (intrepiditee) [reply static/data/hierarchy_statsvar.json]: I dropped the pv tree files from this PR

File: static/js/tools/timeline_chart.tsx (hunk 6 above)
Comment (intrepiditee) [reply static/js/tools/timeline_chart.tsx]: replaced

File: static/js/tools/statsvar_menu.tsx (hunk 5 above)
Comment (intrepiditee) [reply static/js/tools/statsvar_menu.tsx:49]: You are right! It did break the choropleth tool. fixed now

File: server/webdriver_tests/screenshot/screenshot_test.py (hunk 3 above)
Comment (intrepiditee) [reply server/webdriver_tests/screenshot/screenshot_test.py]: removed

File: static/js/tools/timeline_util.ts (hunk 2 above)
Comment (intrepiditee) [reply static/js/tools/timeline_util.ts]: is this better?

File: static/js/tools/statsvar_menu.tsx (hunk 5 above)
Comment (intrepiditee) [reply static/js/tools/statsvar_menu.tsx:49]: Not because of the extra param though. The problem was that the values of StatsVarNodes become objects

**Comment by intrepiditee** [comment]
@shifucun @beets All the comments are addressed

File: static/js/tools/timeline_util.ts (hunk 2 above)
Comment (beets) [reply static/js/tools/timeline_util.ts]: yes, better. thanks!

**Review by beets — APPROVED** [review]
